### PR TITLE
Limit Concurrency for thumb request

### PR DESCRIPTION
**Pull request by revonateB0T**

### CODE REVIEW

**Comment by revonateB0T** [comment]
Note that there is no i18n support for thumb concurrency setting as this option is ought to remove after we find a sane value.

File: app/src/main/kotlin/com/hippo/ehviewer/coil/LimitConcurrencyNetworkClient.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import com.hippo.ehviewer.Settings
+import com.hippo.ehviewer.client.URL_PREFIX_THUMB_E
+import com.hippo.ehviewer.client.URL_PREFIX_THUMB_EX
+import io.ktor.client.HttpClient
+import kotlinx.coroutines.sync.Semaphore
+import kotlinx.coroutines.sync.withPermit
+
+class LimitConcurrencyNetworkClient(val impl: NetworkClient) : NetworkClient {
+    val eSemaphore = Semaphore(Settings.thumbConcurrency)
+    val exSemaphore = Semaphore(Settings.thumbConcurrency)
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/coil/LimitConcurrencyNetworkClient.kt:16]: We are not gonna requesting e and ex at the same time, I think they can share the semaphore.

File: app/src/main/kotlin/com/hippo/ehviewer/coil/LimitConcurrencyNetworkClient.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    override suspend fun <T> executeRequest(req: NetworkRequest, f: suspend (NetworkResponse) -> T): T {
+        val url = req.url
+        return when {
+            URL_PREFIX_THUMB_E in url -> eSemaphore.withPermit { impl.executeRequest(req, f) }
+            URL_PREFIX_THUMB_EX in url -> exSemaphore.withPermit { impl.executeRequest(req, f) }
+            else -> impl.executeRequest(req, f)
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/coil/LimitConcurrencyNetworkClient.kt:22]: What about normal previews?

Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/coil/LimitConcurrencyNetworkClient.kt:22]: They should not be limited as host are not the same, mostly.

Comment (FooIbar) [reply app/src/main/kotlin/com/hippo/ehviewer/coil/LimitConcurrencyNetworkClient.kt:22]: They are the same for the same gallery, so 400 / 20 = 20 parallel requests at most.

File: app/src/main/kotlin/com/hippo/ehviewer/coil/LimitConcurrencyNetworkClient.kt (hunk 1 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/coil/LimitConcurrencyNetworkClient.kt]: Done, Actually ehgt don't need semaphore protect.

**Comment by revonateB0T** [comment]
@FooIbar friendly ping

**Review by FooIbar — APPROVED** [review]
I can't reproduce the issue, but it's fine if it works for you.

**Pull request merged**